src/pip/_internal/metadata/pkg_resources.py

```diff
@@ -63,12 +98,26 @@ def from_wheel(cls, wheel: Wheel, name: str) -> "Distribution":
 
         :raises InvalidWheel: Whenever loading of the wheel causes a
             :py:exc:`zipfile.BadZipFile` exception to be thrown.
+        :raises UnsupportedWheel: If the wheel is a valid zip, but malformed
+            internally.
         """
         try:
             with wheel.as_zipfile() as zf:
-                dist = pkg_resources_distribution_for_wheel(zf, name, wheel.location)
-        except BadZipFile as e:
+                info_dir, _ = parse_wheel(zf, name)
+                metadata_text = {
+                    path.split("/", 1)[-1]: read_wheel_metadata_file(zf, path)
+                    for path in zf.namelist()
+                    if path.startswith(f"{info_dir}/")
+                }
+        except zipfile.BadZipFile as e:
             raise InvalidWheel(wheel.location, name) from e
+        except UnsupportedWheel as e:
+            raise UnsupportedWheel(f"{name} has an invalid wheel, {e}")
```

Comment:
Why not let this error pass through as-is?

The caller has the `name` and can use it, if they really want to, no?

Reply:
Yes, but I don’t really want to think about changing behaviour right now; I’ll get too many chances for that when we try to re-implement everything in importlib.metadata and deal with those subtle differences.